include/frozen/string.h

```diff
@@ -92,191 +88,6 @@ template <> struct elsa<string> {
   }
 };
 
-namespace kmp {
-
-template <std::size_t size> class string_needle {
-  bits::cvector<std::ptrdiff_t, size> cache_;
-  std::array<char, size> data_;
-
-  static constexpr bits::cvector<std::ptrdiff_t, size>
-  build_kmp_cache(std::array<char, size> needle) {
-    char const *needle_view = needle.data();
-    std::ptrdiff_t cnd = 0;
-    bits::cvector<std::ptrdiff_t, size> cache;
-    for (std::size_t pos = 0; pos < size; ++pos)
-      cache[pos] = -1;
-
-    for (std::size_t pos = 1; pos < size; ++pos) {
-      if (needle_view[pos] == needle_view[cnd]) {
-        cache[pos] = cache[cnd];
-        cnd += 1;
-      } else {
-        cache[pos] = cnd;
-        cnd = cache[cnd];
-        while (cnd >= 0 && needle_view[pos] != needle_view[cnd])
-          cnd = cache[cnd];
-        cnd += 1;
-      }
-    }
-    return cache;
-  }
-
-public:
-  constexpr string_needle(std::array<char, size> data)
-      : cache_{build_kmp_cache(data)}, data_(data) {}
-
-  template <std::size_t... I>
-  constexpr string_needle(char const data[size], std::index_sequence<I...>)
-      : string_needle(std::array<char, size>{{data[I]...}}) {}
-  constexpr string_needle(char const data[size])
-      : string_needle(data, std::make_index_sequence<size>()) {}
-
-  constexpr char operator[](std::size_t i) const {
-    return (&std::get<0>(data_))[i];
-  }
-  constexpr std::ptrdiff_t step(std::size_t i) const { return cache_[i]; }
```

Comment:
Any reason why you did not keep this auxillary function (putting it protected if you want)? It makes code reading easier.

Reply:
No particular reason.
What do you think if :
* we change `data_` to be a `bits::cvector` instead of `std::array` ?
* and if `step(i)` just return `cache_[i]`, rename `cache_` to `step_` ?